benchmarks/cpp_benchmark.cc

```diff
@@ -0,0 +1,214 @@
+#define SUFFIX ".pb"
+
+using benchmarks::BenchmarkDataset;
+using google::protobuf::Arena;
+using google::protobuf::Descriptor;
+using google::protobuf::DescriptorPool;
+using google::protobuf::Message;
+using google::protobuf::MessageFactory;
+
+class Fixture : public benchmark::Fixture {
+ public:
+  Fixture(const BenchmarkDataset& dataset, const std::string& suffix) {
+    for (int i = 0; i < dataset.payload_size(); i++) {
+      payloads_.push_back(dataset.payload(i));
+    }
+
+    const Descriptor* d =
+        DescriptorPool::generated_pool()->FindMessageTypeByName(
+            dataset.message_name());
+
+    if (!d) {
+      std::cerr << "Couldn't find message named '" << dataset.message_name()
+                << "\n";
+    }
+
+    prototype_ = MessageFactory::generated_factory()->GetPrototype(d);
+    SetName((dataset.name() + suffix).c_str());
+  }
+
+ protected:
+  std::vector<std::string> payloads_;
+  const Message* prototype_;
+};
+
+class WrappingCounter {
+ public:
+  WrappingCounter(size_t limit) : value_(0), limit_(limit) {}
+
+  size_t Next() {
+    size_t ret = value_;
+    if (++value_ == limit_) {
+      value_ = 0;
+    }
+    return ret;
+  }
+
+ private:
+  size_t value_;
+  size_t limit_;
+};
+
+template <class T>
+class ArenaParseFixture : public Fixture {
+ public:
+  ArenaParseFixture(const BenchmarkDataset& dataset)
+      : Fixture(dataset, "_parse_arena") {}
+
+  virtual void BenchmarkCase(benchmark::State& state) {
+    WrappingCounter i(payloads_.size());
+    size_t total = 0;
+
+    while (state.KeepRunning()) {
+      Arena arena;
+      Message* m = Arena::CreateMessage<T>(&arena);
+      const std::string& payload = payloads_[i.Next()];
+      total += payload.size();
+      m->ParseFromString(payload);
+    }
+
+    state.SetBytesProcessed(total);
+  }
+};
+
+class NoArenaParseFixture : public Fixture {
+ public:
+  NoArenaParseFixture(const BenchmarkDataset& dataset)
+      : Fixture(dataset, "_parse_noarena") {}
+
+  virtual void BenchmarkCase(benchmark::State& state) {
+    Message* m = prototype_->New();
+    WrappingCounter i(payloads_.size());
+    size_t total = 0;
+
+    while (state.KeepRunning()) {
+      const std::string& payload = payloads_[i.Next()];
+      total += payload.size();
+      m->ParseFromString(payload);
+    }
+
+    state.SetBytesProcessed(total);
+    delete m;
+  }
+};
+
+class SerializeFixture : public Fixture {
+ public:
+  SerializeFixture(const BenchmarkDataset& dataset)
+      : Fixture(dataset, "_serialize") {}
+
+  virtual void BenchmarkCase(benchmark::State& state) {
+    std::vector<Message*> messages;
+    for (size_t i = 0; i < payloads_.size(); i++) {
+      messages.push_back(prototype_->New());
+      messages.back()->ParseFromString(payloads_[i]);
+    }
+
+    size_t total = 0;
+    std::string str;
+    WrappingCounter i(payloads_.size());
+
+    while (state.KeepRunning()) {
+      str.clear();
+      messages[i.Next()]->SerializeToString(&str);
```

Comment:
Allocation should only happen for the first time through the loop. `str.clear()` won't release the memory.